mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala

```diff
@@ -234,3 +249,298 @@ object SMTEncoding {
+
+/**
+ * Specifies the options used to control pass executions
+ *
+ * Typically, each pass will need a subset of the available option groups, and the sequence of passes run by the
+ * `PassExecutor` will require the union of the option groups required by its constituent passes.
+ *
+ * The unions can be specified via the `Has*` traits, such as [[OptionGroup.HasInput]] and constructed via the `With*`
+ * classes, such as [[OptionGroup.WithInput]].
+ */
+object OptionGroup {
+
+  /**
+   * Interface for a group of related options that can be be produced from a [[Config]]
+   *
+   * The intended use of this class is to identify '''configurable''' options. Configurable options are extensions of
+   * `OptionGroup` that can be derived from a `Config`. Typically, these are case classes.
+   *
+   * @param cfg
+   *   An instance of [[Config config group]].
+   */
+  sealed trait Configurable[C <: Config[C], O] {
+    //  TODO could manual apply methods be replaced with pureconfig merging?
+    def apply(cfg: C): Try[O]
+  }
+
+  // Convert optional values into `Try`'s
+  // see https://stackoverflow.com/questions/17521709/how-can-i-best-convert-an-option-into-a-try/45017589#45017589
+  implicit class OptionOps[A](opt: Option[A]) {
+
+    def toTry(field: String): Try[A] = {
+      opt
+        .map(Success(_))
+        .getOrElse(Failure(new PassOptionException(s"Missing value for required option ${field}")))
+    }
+  }
+
+  /** Options used in all modes of execution */
+  case class Common(
+      command: String,
+      debug: Boolean,
+      features: Seq[Feature],
+      outDir: File,
+      profiling: Boolean,
+      runDir: Option[File],
+      smtprof: Boolean,
+      writeIntermediate: Boolean)
+      extends OptionGroup
+
+  object Common extends Configurable[Config.Common, Common] {
+    // NOTE: These conversions can probably be automated via some
+    // clever use of shapeless records, but not sure if it's worth the
+    // complexity at this point.
+    //
+    // If we change the needed options so that all values are non `Option`, the
+    // automated conversion would be trivially.
+    //
+    def apply(common: Config.Common): Try[Common] = {
+      for {
+        // Required fields
+        command <- common.command.toTry("common.command")
+        debug <- common.debug.toTry("debug")
+        features <- common.features.toTry("common.features")
+        outDir <- common.outDir.toTry("common.outDir")
+        profiling <- common.profiling.toTry("common.profiling")
+        smtprof <- common.debug.toTry("common.smtprog")
+        writeIntermediate <- common.writeIntermediate.toTry("common.writeIntermediate")
+      } yield Common(
+          command = command,
+          debug = debug,
+          features = features,
+          outDir = outDir,
+          profiling = profiling,
+          runDir = common.runDir, // Remains optional
+          smtprof = smtprof,
+          writeIntermediate = writeIntermediate,
+      )
+    }
+  }
```

Comment:
I'm not pleased with how much typing was involved in all this, but I couldn't think of a cleaner way to get the desired guarantees (i.e., of a fully configured set of options for passes) without doing some real non-trivial type astronaut stuff with shapeless.

I'm open to other ideas tho!

The upside is that at least the logic here is dead simple.